File: materials/metal.py
from materials.material import *

class Metal(Material):
    def __init__(self, color, fuzz=0):
        super().__init__(color)
        self.fuzz = fuzz

    def bounce(self, r, normal, hit_pos):
        new_direction = r.direction - 2 * r.direction.dot(normal)*normal + self.fuzz * Vector.random_in_unit_sphere()
        return Ray(hit_pos, new_direction)

File: materials/metal.cpp
#include "metal.h"

#include "../ray.h"

Metal::Metal(const Vector &color, float fuzz) : Material(color), fuzz(fuzz) {}

Ray Metal::bounce(const Ray &r, const Vector &normal, const Vector &hitPos) const {
    Vector reflected = r.direction.normalize() - 2 * r.direction.normalize().dot(normal) * normal;
    if (fuzz == 0)
    {
        return Ray(hitPos, reflected);
    }
    
    return Ray(hitPos, reflected + fuzz * Vector::randomInUnitSphere());
}